urld829b80034_wjch111_PTSP2013Wang_tgz-pJ8-controllers_momcts_momctsCore_TestPlannerJ8/src/controllers/momcts/utils/Transformation.java

```diff
@@ -0,0 +1,74 @@
+package controllers.momcts.utils;
+
+import java.util.HashMap;
+import java.util.Vector;
+
+/**
+ * This class contains all frequently used type transformations of Weijia WANG programs
+ * @author wj
+ * @date 2013/07/11
+ */
+public class Transformation {
+	public static Vector<String> ints2strs(Vector<Integer> ints){
+		Vector<String> strs = new Vector<String>();
+		for(int i=0; i< ints.size();i++){
+			strs.add(""+ints.get(i));
+		}
+		return strs;
+	}
+	
+	public static Vector<Integer> intAry2Vec(int[] is){
+		Vector<Integer> ints = new Vector<Integer>();
+		for(int i=0;i<is.length;i++){
+			ints.add(is[i]);
+		}
+		return ints;
+	}
+	
+	public static Vector<Double> doubleAry2Vec(double[] ds){
+		Vector<Double> dubs = new Vector<Double>();
+		for(int i=0;i<ds.length;i++){
+			dubs.add(ds[i]);
+		}
+		return dubs;	
+	}
+
+	public static Vector<String> strAry2Vec(String[] ss){
+		Vector<String> strs = new Vector<String>();
+		for(int i=0;i<ss.length;i++){
+			strs.add(ss[i]);
+		}
+		return strs;
+	}
+	
+	public static Vector<Double> ints2doubles(int[] is){
+		Vector<Double> ds = new Vector<Double>();
+		for(int i:is){
+			double d = i;
+			ds.add(d);
+		}
+		return ds;
+	}
+
+	
+	/**
+	 * Given the keys and values, return a correspondent map
+	 * @param keys
+	 * @param values
+	 * @assure	keys and values have the same size
+	 * @return
+	 */
+	public static <K,V> HashMap<K,V> vec2map(Vector<K> keys, Vector<V> values){
+		HashMap<K,V> mp = new HashMap<K,V>();
+		int vlSize = values.size();
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem:**
 This line of code lacks validation when processing input data through the following parameter: &#39;values&#39; (index: 1 | type: Vector). The parameter is exposed to external callers, because its enclosing class and method are publicly accessible. This means that upstream validation, if it exists, can be bypassed. Other validated parameters: &#39;keys&#39;. The same parameter type is validated here for example: urld829b80034_wjch111_PTSP2013Wang_tgz-pJ8-controllers_momcts_momctsCore_TestPlannerJ8/src/controllers/momcts/MOOTools.java:29. Malicious, malformed, or unbounded inputs can cause unexpected runtime behavior or crashes, and can slow performance.

**Fix:**
 Add checks to ensure the validity of the parameter&#39;s value, such as testing it for nullness (for example, using the @NonNull annotation as [described in the Lombok library](https://projectlombok.org/features/NonNull)), emptiness, or equality. Or to prevent direct calls to it, [reduce the method&#39;s visibility](https://docs.oracle.com/javase/tutorial/java/javaOO/accesscontrol.html).

 Learn more about potential threats and guidance from the [Common Weakness Enumeration website](https://cwe.mitre.org/data/definitions/693.html) and the [OWASP Cheat Sheet series](https://cheatsheetseries.owasp.org/cheatsheets/Input_Validation_Cheat_Sheet.html).